napi-inl.h

```diff
@@ -3106,11 +3137,11 @@ inline ObjectWrap<T>::ObjectWrap(const Napi::CallbackInfo& callbackInfo) {
   napi_value wrapper = callbackInfo.This();
   napi_status status;
   napi_ref ref;
-  T* instance = static_cast<T*>(this);
```

Comment:
@dmitryash it's actually possible to avoid this cast, because `this` works just fine in the code below.

Reply:
@gabrielschulhof While I agree that this is nicer this way, it also requires updating the cast in `FinalizeCallback` like #475 does

Reply:
@addaleax I think  #475 will mostly supersede this PR.